nexus/types/src/deployment/planning_input.rs

```diff
@@ -457,12 +457,31 @@ pub enum SledFilter {
     // ---
     // Prefer to keep this list in alphabetical order.
     // ---
+    /// All sleds that the blueprint execution background task should attempt to
```

Comment:
I had an immediate negative reaction to this structure, but I think I'm coming around to it. Let me try to articulate "what's weird about this enum" to me.

# Two approaches

I think there's a decision whether we want this enum to describe **tuples of state/policy**, or whether we want this enum to describe **use-cases**, which can be matched against state and policy. I believe there is a distinction, but I think it's subtle.

- For example: the usage of filters like `Commissioned`, `InService`, `Decommissioned`, imply **description about state and policy**. These don't really describe "what action is being taken", or "what summation of sleds" we want, but rather, they sorta try to describe the state/policy pair.
- In contrast: filters like `CouldBeRunningInstances`, `EligibleForZpools`, `OperatorVisible` describe use-cases, where the exact policy/state matches are covered by the match cases, and could plausibly change as we update the system.

# Maybe we do one approach? Maybe we make this distinction more clear?

I think the combination of both of these together is a bit weird in `SledFilter`, because as a person-writing-code, it's not clear which pathway I should take. If I have a new use-case that happens to align with `Commissioned`, should I use that variant? Or should I add a new variant with a name that describes what I'm doing?

All this is to say: I personally think it would be less confusing if we went all-in on the "summarize tuples" or the "go by use-cases" approach, but didn't mix them.

- If we go down the "summarize tuples" path, presumably we'd try to enumerate the desirable pairings of (state,policy), and give each pair a unique name that we use.
- If we go down the "use-case" path, presumably we would not care about overlap, we'd just keep adding tuples for each new use-case.

# Does this make sense?

From what I can tell, it seems like you're preferring the "go by use-case" approach -- for example, `CouldBeRunningDiscretionaryServices` and `CouldBeRunningInstances` have perfect overlap, but you seemed to want to split them explicitly. 

I think the `Commissioned` and `Decommissioned` are still "summarizing the tuple", but all other variants describe use-cases. I'm not really advocating for getting rid of them, but do we want to make them "more clearly distinct" from the other use-cases somehow? (Notably both of these just ignore policy)

Reply:
Thanks for the summary, Sean! I was definitely trying to go for the use-cases approach, under the theory that as we continue work on the system and as we add new states (particularly around graceful removal), our understanding of which states should count for which uses will change over time.

However: maybe we need to reconsider this! I definitely hadn't anticipated quite these many variants, and the number here is concerning. It's probably worth thinking about all the places we use sled filters and seeing which approach would make more sense. For example, maybe we flip the cost around instead: rather than adding a new variant for each use case, we inspect all of Nexus (perhaps with the help of rust-analyzer) when we add a new state or policy. In this alternate world, part of adding a new state would be to look at every current use of `SledFilter` and see what to do.

I guess one aspiration could be to have a variant for every group of use cases that, when a new state or policy is added, are reasonably likely to have similar behavior. But that seems very difficult to achieve.

Reply:
At this moment I'm pretty strongly in favor of the use-cases approach. I'm not super concerned about the growth of enum variants, and it will be much easier to scale to larger numbers of developers if they can see exactly what they need for their service. They can then ask someone if they need a new one. I don't see how the look at each existing use of `SledFilter` every time will scale, and it seems like a tighter dev process than we usually use. By tighter, I mean, the variants will be much more "controlled".

I think the main reason I favor the use-cases approach though is that putting a combined variant for each tuple combination (at least valid ones), makes it feel like state and policy could almost be the same thing. I know they can't as they are used elsewhere for different purposes, but I feel like trying to compress them down to a single variant for each use is actually less easy than the alternative. It's also much less clear at the call site. Maybe this is a silly concern, but it's something I felt.